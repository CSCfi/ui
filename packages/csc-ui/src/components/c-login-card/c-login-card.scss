:host {
  display: flex;
}

::slotted(*) {
  max-width: 70%;
}

.c-login-card {
  border-radius: 6px;
  box-shadow: rgb(0 0 0 / 15%) 0px 10px 20px, rgb(0 0 0 / 20%) 0px 5px 5px;
  display: flex;
  position: relative;
  width: 100%;

  &__image {
    background-size: cover;
    border-radius: 6px;
    clip-path: url(#myClip);
    height: 100%;
    left: 0;
    overflow: hidden;
    position: absolute;
    top: 0;
    width: 100%;

    &--overlay {
      background-color: var(--csc-primary);
      background-blend-mode: var(--c-login-overlay-mode);
    }
  }

  &__content {
    display: flex;
    flex-direction: column;
    gap: 24px;
    padding: 72px;
    margin-bottom: 32px;
  }

  &--mobile {
    flex-direction: column;

    ::slotted(*) {
      max-width: 100%;
    }

    ::slotted(c-login-card-title) {
      font-size: 32px;
      hyphens: auto;
    }

    .c-login-card__content {
      background-color: #fff;
      border-radius: 6px;
      margin: -6px 0 0;
      padding: 32px;
      z-index: 1;
    }

    .c-login-card__image {
      aspect-ratio: 16 / 9;
      border-radius: 6px 6px 0 0;
      box-shadow: inset 0 -16px 16px -18px #000;
      clip-path: none;
      min-height: 144px;
      position: relative;
      width: 100%;
    }
  }
}
