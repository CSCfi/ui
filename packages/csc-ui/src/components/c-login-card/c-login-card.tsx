import { Component, Element, Host, h, Prop } from '@stencil/core';

export type CLoginCardBlendMode =
  | 'normal'
  | 'multiply'
  | 'screen'
  | 'overlay'
  | 'darken'
  | 'lighten'
  | 'color-dodge'
  | 'color-burn'
  | 'hard-light'
  | 'soft-light'
  | 'difference'
  | 'exclusion'
  | 'hue'
  | 'saturation'
  | 'color'
  | 'luminosity';

/**
 * @group Cards
 * @slot - Login Card contents
 */
@Component({
  tag: 'c-login-card',
  styleUrl: 'c-login-card.scss',
  shadow: true,
})
export class CLoginCard {
  /**
   * Background position (css background-position)
   */
  @Prop() backgroundPosition = 'bottom right';

  /**
   * Mobile breakpoint in pixels
   */
  @Prop() mobileBreakpoint = 600;

  /**
   * Add colored overlay to the background image
   */
  @Prop() overlay = false;

  /**
   * Add colored overlay to the background image
   */
  @Prop() overlayBlendMode: CLoginCardBlendMode = 'multiply';

  /**
   * Background image
   */
  @Prop() src = '';

  @Element() element: HTMLCLoginCardElement;

  private _observer: ResizeObserver;

  private _cardElement: HTMLDivElement;

  componentDidLoad() {
    this._observer = new ResizeObserver(([entry]) => {
      const { width } = entry.contentRect;

      this._cardElement.classList.toggle(
        'c-login-card--mobile',
        width <= this.mobileBreakpoint,
      );
    });

    this._observer.observe(this.element);
  }

  disconnectedCallback() {
    this._observer.disconnect();
  }

  render() {
    const style = {
      backgroundImage: `url(${this.src})`,
      backgroundPosition: this.backgroundPosition,
      '--c-login-overlay-mode': !!this.overlay && this.overlayBlendMode,
    };

    const imageClasses = {
      'c-login-card__image': true,
      'c-login-card__image--overlay': !!this.overlay,
    };

    return (
      <Host>
        <svg width="0" height="0">
          <clipPath id="myClip" clipPathUnits="objectBoundingBox">
            <path d="m0.234,0.914 c0.132,-0.026,0.286,-0.05,0.436,-0.163 c0.083,-0.063,0.152,-0.145,0.21,-0.329 c0.055,-0.172,0.072,-0.421,0.072,-0.421 h0.048 v1 h-1 v-0.057 c0,0,0.145,-0.012,0.234,-0.029" />
          </clipPath>
        </svg>

        <div
          class="c-login-card"
          ref={(el) => (this._cardElement = el as HTMLDivElement)}
        >
          <div class={imageClasses} style={style}></div>

          <div class="c-login-card__content">
            <slot></slot>
          </div>
        </div>
      </Host>
    );
  }
}
