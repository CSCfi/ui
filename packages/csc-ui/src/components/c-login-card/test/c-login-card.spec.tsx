import { newSpecPage } from '@stencil/core/testing';
import { CLoginCard } from '../c-login-card';

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

describe('c-login-card', () => {
  const content = `This is a login card`;
  it('renders default html', async () => {
    const page = await newSpecPage({
      components: [CLoginCard],
      html: `<c-login-card>
      ${content}
      </c-login-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('add source to background-image', async () => {
    const page = await newSpecPage({
      components: [CLoginCard],
      html: `<c-login-card src="mahti">
      ${content}
      </c-login-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders custom background-position', async () => {
    const page = await newSpecPage({
      components: [CLoginCard],
      html: `<c-login-card background-position="top left" src="mahti">
      ${content}
      </c-login-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders custom background-position', async () => {
    const page = await newSpecPage({
      components: [CLoginCard],
      html: `<c-login-card background-position="25% 50%" src="mahti">
      ${content}
      </c-login-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders custom breakpoint for mobile', async () => {
    const page = await newSpecPage({
      components: [CLoginCard],
      html: `<c-login-card mobile-breakpoint="500">
      ${content}
      </c-login-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders colored overlay', async () => {
    const page = await newSpecPage({
      components: [CLoginCard],
      html: `<c-login-card overlay=${true} 
      overlay-blend-mode="hue"  
      background-position="top left" 
      src="mahti">
      ${content}
      </c-login-card>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
