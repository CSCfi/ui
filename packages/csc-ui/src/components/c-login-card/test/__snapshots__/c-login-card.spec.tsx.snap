// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-login-card add source to background-image 1`] = `
<c-login-card src="mahti">
  <!---->
  <svg height="0" width="0">
    <defs>
      <clipPath clipPathUnits="objectBoundingBox" id="cLoginClipPath">
        <path d="m1,1 h-1 v-0.213 c0,0,0.209,-0.046,0.337,-0.109 c0.191,-0.096,0.413,-0.183,0.629,-0.608 c0.028,-0.055,0.034,-0.069,0.034,-0.069"></path>
      </clipPath>
    </defs>
  </svg>
  <div class="c-login-card c-login-card--mobile">
    <div class="c-login-card__image" style="background-image: url(mahti); background-position: bottom right; height: 0px; --c-login-overlay-mode: false;"></div>
    <div class="c-login-card__content">
      This is a login card
    </div>
  </div>
</c-login-card>
`;

exports[`c-login-card renders colored overlay 1`] = `
<c-login-card background-position="top left" overlay="true" overlay-blend-mode="hue" src="mahti">
  <!---->
  <svg height="0" width="0">
    <defs>
      <clipPath clipPathUnits="objectBoundingBox" id="cLoginClipPath">
        <path d="m1,1 h-1 v-0.213 c0,0,0.209,-0.046,0.337,-0.109 c0.191,-0.096,0.413,-0.183,0.629,-0.608 c0.028,-0.055,0.034,-0.069,0.034,-0.069"></path>
      </clipPath>
    </defs>
  </svg>
  <div class="c-login-card c-login-card--mobile">
    <div class="c-login-card__image c-login-card__image--overlay" style="background-image: url(mahti); background-position: top left; height: 0px; --c-login-overlay-mode: hue;"></div>
    <div class="c-login-card__content">
      This is a login card
    </div>
  </div>
</c-login-card>
`;

exports[`c-login-card renders custom background-position 1`] = `
<c-login-card background-position="top left" src="mahti">
  <!---->
  <svg height="0" width="0">
    <defs>
      <clipPath clipPathUnits="objectBoundingBox" id="cLoginClipPath">
        <path d="m1,1 h-1 v-0.213 c0,0,0.209,-0.046,0.337,-0.109 c0.191,-0.096,0.413,-0.183,0.629,-0.608 c0.028,-0.055,0.034,-0.069,0.034,-0.069"></path>
      </clipPath>
    </defs>
  </svg>
  <div class="c-login-card c-login-card--mobile">
    <div class="c-login-card__image" style="background-image: url(mahti); background-position: top left; height: 0px; --c-login-overlay-mode: false;"></div>
    <div class="c-login-card__content">
      This is a login card
    </div>
  </div>
</c-login-card>
`;

exports[`c-login-card renders custom background-position 2`] = `
<c-login-card background-position="25% 50%" src="mahti">
  <!---->
  <svg height="0" width="0">
    <defs>
      <clipPath clipPathUnits="objectBoundingBox" id="cLoginClipPath">
        <path d="m1,1 h-1 v-0.213 c0,0,0.209,-0.046,0.337,-0.109 c0.191,-0.096,0.413,-0.183,0.629,-0.608 c0.028,-0.055,0.034,-0.069,0.034,-0.069"></path>
      </clipPath>
    </defs>
  </svg>
  <div class="c-login-card c-login-card--mobile">
    <div class="c-login-card__image" style="background-image: url(mahti); background-position: 25% 50%; height: 0px; --c-login-overlay-mode: false;"></div>
    <div class="c-login-card__content">
      This is a login card
    </div>
  </div>
</c-login-card>
`;

exports[`c-login-card renders custom breakpoint for mobile 1`] = `
<c-login-card mobile-breakpoint="500">
  <!---->
  <div class="c-login-card c-login-card--mobile">
    <div class="c-login-card__content c-login-card__content--no-image">
      This is a login card
    </div>
  </div>
</c-login-card>
`;

exports[`c-login-card renders default html 1`] = `
<c-login-card>
  <!---->
  <div class="c-login-card c-login-card--mobile">
    <div class="c-login-card__content c-login-card__content--no-image">
      This is a login card
    </div>
  </div>
</c-login-card>
`;
